flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/interval/SemiAntiTimeIntervalJoin.java

```diff
@@ -0,0 +1,450 @@
+    private final long stateRetentionTime;
+
+    // state to store rows from the left stream
+    private transient SemiAntiLeftRecordStateView leftRecordStateView;
+    // state to store rows from the right stream
+    private transient SemiAntiRightRecordStateView rightRecordStateView;
+
+    // Current time on the respective input stream.
+    protected long leftOperatorTime = 0L;
+    protected long rightOperatorTime = 0L;
+
+    public SemiAntiTimeIntervalJoin(
+            boolean isAntiJoin,
+            long leftLowerBound,
+            long leftUpperBound,
+            long allowedLateness,
+            InternalTypeInfo<RowData> leftType,
+            InternalTypeInfo<RowData> rightType,
+            IntervalJoinFunction joinFunction,
+            long stateRetentionTime) {
+        this.isAntiJoin = isAntiJoin;
+        this.leftRelativeSize = -leftLowerBound;
+        this.rightRelativeSize = leftUpperBound;
+        if (allowedLateness < 0) {
+            throw new IllegalArgumentException("The allowed lateness must be non-negative.");
+        }
+        this.allowedLateness = allowedLateness;
+        this.leftType = leftType;
+        this.rightType = rightType;
+        this.joinFunction = joinFunction;
+        this.stateRetentionTime = stateRetentionTime;
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        joinFunction.setRuntimeContext(getRuntimeContext());
+        joinFunction.open(parameters);
+
+        // TODO: support joins with unique keys
+        // now states only support no unique key
+        this.leftRecordStateView =
+                new SemiAntiLeftRecordStateView(
+                        getRuntimeContext(),
+                        LEFT_RECORDS_STATE_NAME,
+                        leftType,
+                        createTtlConfig(stateRetentionTime));
+
+        this.rightRecordStateView =
+                new SemiAntiRightRecordStateView(
+                        getRuntimeContext(),
+                        RIGHT_RECORDS_STATE_NAME,
+                        rightType,
+                        createTtlConfig(stateRetentionTime));
+    }
+
+    @Override
+    public void processElement1(RowData leftRow, Context ctx, Collector<RowData> out)
+            throws Exception {
+        updateOperatorTime(ctx);
+
+        long timeForLeftRow = getTimeForLeftStream(ctx, leftRow);
+        long rightQualifiedLowerBound = timeForLeftRow - rightRelativeSize;
+        long rightQualifiedUpperBound = timeForLeftRow + leftRelativeSize;
+
+        List<RowDataInfo> associatedRecords =
+                getRowDataAssociatedWithOtherInput(
+                        leftRow,
+                        true,
+                        joinFunction.getJoinCondition(),
+                        rightQualifiedLowerBound,
+                        rightQualifiedUpperBound);
+
+        if (associatedRecords.isEmpty()) {
```

Comment:
For semiJoin, we could emit result per record.
However for antiJoin, we could only emit result when right side watermark exceeds registered clear timer. Right?

Reply:
I think what you say is also a good idea. Currently in my code, anti join have the same logic to emit result per record just like semi join. However, when meet a record and should retract one record we emit before, a record -D will be emit to keep the correct semantics.

Reply:
Your proposal is reasonable, too. However, if we use this way, the output stream is antiJoin an update stream instead of append stream, which is different with other interval join type.

Reply:
So, it just likes a mini batch and the batch size is the window size ? In my immature vision, interval join maybe should be an retract stream to conform to the semantics of the stream. Otherwise if the user defines a month-level window size, it will produce no results until the window end time arrives.

Reply:
In current implementation, interval join only support consuming insert-only and producing insert-only changes. We depend this consumption when infer `ChangelogMode` for every physical node, you could find more detailed in `FlinkChangelogModeInferenceProgram`.
IMO, It's better to keep consistent behavior for a same operator, no matter it's left/right/full/inner/semi/anti join type.
If you think this behavior is not reasonable. We could drive a discussion in dev mail list. If more users and developers prefer to emit result intermediately, we could update the behavior for all join type of interval join, not only anti join.
WDYT?

Reply:
agree

Reply:
So maybe it's better to also let semi join keep the same behavior with anti join that emit results when the expired time arrived instead of emitting result per record. Because I think when user use these two kinds of join, and find they have the different behaviors, he will be confused. What do you think?

Reply:
Currently, interval join requires both left input and right input are append stream. So I think it's ok to emit result by per input record for semiJoin. When one left input record comes, it could emit result it find any matched records in right side. When right input record comes, it would check if it matched records in left side and left side record has never emit before, then left side could emit to downstream.

Maybe it would not confused user, because existed left/right/full outer join apply this behavior. You could tell whether a record is matched with the other side earlier, because you could only tell whether a record could never matched the other side until the last time in allowed time ranges, right?

Reply:
Thanks for your explanation sincerely, I'll fix it.

Reply:
This class is deleted and the code about semi/anti will be merged into the current TimeIntervalJoin.java.